src/main/java/com/itechart/students_lab/waybill_suppliers/service/WarehouseService.java

```
@@ -67,17 +64,19 @@ public WarehouseDto findById(Long id) {
 
     public WarehouseDto create(WarehouseDto warehouseDto) {
         Warehouse warehouse = warehouseMapper.warehouseDtoToWarehouse(warehouseDto);
-        Optional<Customer> customer = customerRepo.findById(warehouseDto.getCustomerId());
-        if (customer.isPresent()) {
-            if (customer.get().getActiveStatus() == ActiveStatus.INACTIVE) {
-                throw new ServiceException(HttpStatus.CONFLICT,
-                        FAILED_CREATE_WAREHOUSE_CUSTOMER_DEACTIVATED);
-            }
-            warehouse.setCustomer(customer.get());
-        } else {
-            throw new EntityNotFoundException(
-                    String.format(CUSTOMER_WITH_ID_NOT_FOUND, warehouseDto.getCustomerId()));
+        Customer customer = customerService.getActiveCustomer(warehouseDto.getCustomerId());
+
+        Address address = warehouse.getAddress();
+        if (!warehouseRepo.findByAddress(
+                address.getState(),
+                address.getCity(),
+                address.getFirstAddressLine(),
+                address.getSecondAddressLine()
+        ).isEmpty()) {
```

just .orElseThrow instead of redundant if with body

---

the return value of the method is a list of warehouses, I did it because despite all the constraints in db, it's still possible to manually add the warehouse with the same address but with different id. In that situation findByAdress() method will return 2 warehouses (while 1 is expected, if the return value is Optional), what will cause a server error. So, if there's no need to handle such situations, I'll replace it with Optional, but anyway orElseThrow() method is invoked when optional is empty, while I need to throw an exception when it's not.

---

What do you mean by "manually add"?
In any case, then this is a problem of constraint, you can tune it to refuse duplicates. In here, make a composite constraint on address city and address lines, it should work